sdk/packet/src/lib.rs

```diff
@@ -158,11 +161,76 @@ impl Packet {
         &mut self.meta
     }
 
+    #[cfg(feature = "bincode")]
+    /// Initializes a std::mem::MaybeUninit<Packet> such that the Packet can
+    /// be safely extracted via methods such as MaybeUninit::assume_init()
+    pub fn init_packet_from_data<T: serde::Serialize>(
+        packet: &mut mem::MaybeUninit<Packet>,
+        data: &T,
+        addr: Option<&SocketAddr>,
+    ) -> Result<()> {
+        let mut writer = PacketWriter::new_from_uninit_packet(packet);
+        bincode::serialize_into(&mut writer, data)?;
+
+        let serialized_size = writer.position();
+        let (ip, port) = if let Some(addr) = addr {
+            (addr.ip(), addr.port())
+        } else {
+            (IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0)
+        };
+        Self::init_packet_meta(
+            packet,
+            Meta {
+                size: serialized_size,
+                addr: ip,
+                port,
+                flags: PacketFlags::empty(),
+            },
+        );
+
+        Ok(())
+    }
+
+    pub fn init_packet_from_bytes(
+        packet: &mut mem::MaybeUninit<Packet>,
+        bytes: &[u8],
+        addr: Option<&SocketAddr>,
+    ) -> io::Result<()> {
+        let mut writer = PacketWriter::new_from_uninit_packet(packet);
+        let num_bytes_written = writer.write(bytes)?;
```

Comment:
This should probably use `Write::write_all`.
`Write::write` is not meant to write the entire buffer.
And there is no need to rely on the implementation details of `PacketWriter`.

Reply:
> And there is no need to rely on the implementation details of PacketWriter.

Yeah, that's fair and also simplifies things on the caller side (ie no longer need the `debug_assert`); will make this change